src/Build.UnitTests/Evaluation/Expander_Tests.cs

```diff
@@ -3969,6 +3969,49 @@ public void ExpandItemVectorFunctions_GetPathsOfAllDirectoriesAbove()
             }
         }
 
+        [Fact]
+        public void ExpandItemVectorFunctions_GetPathsOfAllDirectoriesAbove_ReturnCanonicalPaths()
+        {
+            // Directory structure:
+            // <temp>\
+            //    alpha\
+            //        .proj
+            //        One.cs
+            //        beta\
+            //            Two.cs
+            //    gamma\
+            //        Three.cs
+            using (var env = TestEnvironment.Create())
+            {
+                var root = env.CreateFolder();
+
+                var alpha = root.CreateDirectory("alpha");
+                var projectFile = env.CreateFile(alpha, ".proj",
+                    @"<Project>
+  <ItemGroup>
+    <Compile Include=""One.cs"" />
+    <Compile Include=""beta\Two.cs"" />
+    <Compile Include=""..\gamma\Three.cs"" />
+  </ItemGroup>
+  <ItemGroup>
+    <MyDirectories Include=""@(Compile->GetPathsOfAllDirectoriesAbove())"" />
+  </ItemGroup>
+</Project>");
+
+                var beta = alpha.CreateDirectory("beta");
+                var gamma = root.CreateDirectory("gamma");
+
+                ProjectInstance projectInstance = new ProjectInstance(projectFile.Path);
+                ICollection<ProjectItemInstance> myDirectories = projectInstance.GetItems("MyDirectories");
+
+                var includes = myDirectories.Select(i => i.EvaluatedInclude);
+                includes.ShouldContain(root.Path);
+                includes.ShouldContain(alpha.Path);
+                includes.ShouldContain(beta.Path);
+                includes.ShouldContain(gamma.Path);
```

Comment:
Also a `includes.Count.ShouldBe(4);` or similar, to make sure things aren't duplicated?

Reply:
The problem is I don't know how many paths there should be, as that depends on how deeply nested the temp directory is.

Reply:
I think a combination of `ShouldContain` and `ShouldBeUnique` is the right way to go.

Reply:
Done.